File: app/src/main/java/com/gongzelong/iflixtask/Pi.java
package com.gongzelong.iflixtask;

import java.math.BigDecimal;

import static java.math.BigDecimal.ONE;
import static java.math.BigDecimal.ROUND_HALF_UP;

class Pi {
    private OnPiChangedInterface mOnPiChangedInterface;

    interface OnPiChangedInterface {
        void onPiChanged(String pi);
    }

    private static final BigDecimal TWO = new BigDecimal(2);
    private static final BigDecimal FOUR = new BigDecimal(4);
    private String mNumber;

    void pi(final int scale) {
        BigDecimal a = ONE;
        BigDecimal b = ONE.divide(sqrt(TWO, scale), scale, ROUND_HALF_UP);
        BigDecimal t = new BigDecimal(0.25);
        BigDecimal x = ONE;
        BigDecimal y;

        while (!a.equals(b)) {
            y = a;
            a = a.add(b).divide(TWO, scale, ROUND_HALF_UP);
            b = sqrt(b.multiply(y), scale);
            t = t.subtract(x.multiply(y.subtract(a).multiply(y.subtract(a))));
            x = x.multiply(TWO);
        }

        BigDecimal pi = a.add(b).multiply(a.add(b)).divide(t.multiply(FOUR), scale, ROUND_HALF_UP);
        number(pi);

        // print the output of pi result
        if (mOnPiChangedInterface != null) {
            mOnPiChangedInterface.onPiChanged(pi.toString());
        }
    }

    void setOnPiChangedInterface(OnPiChangedInterface onPiChangedInterface) {
        mOnPiChangedInterface = onPiChangedInterface;
    }

    private void number(BigDecimal a) {
        mNumber = a.toString();
    }

    String returnNum() {
        return mNumber;
    }

    private static BigDecimal sqrt(BigDecimal A, final int SCALE) {
        BigDecimal x0 = new BigDecimal("0");
        BigDecimal x1 = new BigDecimal(Math.sqrt(A.doubleValue()));

        while (!x0.equals(x1)) {
            x0 = x1;
            x1 = A.divide(x0, SCALE, ROUND_HALF_UP);
            x1 = x1.add(x0);
            x1 = x1.divide(TWO, SCALE, ROUND_HALF_UP);
        }

        return x1;
    }
}

File: app/src/main/java/com/gongzelong/iflixtask/Pi.kt
package com.gongzelong.iflixtask

import java.math.BigDecimal

import java.math.BigDecimal.ONE
import java.math.BigDecimal.ROUND_HALF_UP

internal class Pi {
    private var mNumber: String? = null
    fun pi(scale: Int) {
        var a = ONE
        var b = ONE.divide(sqrt(TWO, scale), scale, ROUND_HALF_UP)
        var t = BigDecimal(0.25)
        var x = ONE
        var y: BigDecimal

        while (a != b) {
            y = a
            a = a.add(b).divide(TWO, scale, ROUND_HALF_UP)
            b = sqrt(b.multiply(y), scale)
            t = t.subtract(x.multiply(y.subtract(a).multiply(y.subtract(a))))
            x = x.multiply(TWO)
        }

        val pi = a.add(b).multiply(a.add(b)).divide(t.multiply(FOUR), scale, ROUND_HALF_UP)
        number(pi)
    }

    private fun number(a: BigDecimal) {
        mNumber = a.toString()
    }

    fun returnNum(): String? {
        return mNumber
    }

    companion object {

        private val TWO = BigDecimal(2)
        private val FOUR = BigDecimal(4)

        private fun sqrt(A: BigDecimal, SCALE: Int): BigDecimal {
            var x0 = BigDecimal("0")
            var x1 = BigDecimal(Math.sqrt(A.toDouble()))

            while (x0 != x1) {
                x0 = x1
                x1 = A.divide(x0, SCALE, ROUND_HALF_UP)
                x1 = x1.add(x0)
                x1 = x1.divide(TWO, SCALE, ROUND_HALF_UP)
            }

            return x1
        }
    }
}